### ISSUE TITLE
Processing requests to different models in parallel is slower than sequentially

### ISSUE THREAD

**Issue by sermakarevich** [pytorch/serve#481]
Hi, this is a question, not a bug report.

We have torchserve deployed with 5 different models, 1 worker for every model, instance has 12 threads . The goal is to query all models at the same time.

When we do this with python multiprocessing.Pool - it takes multiple times longer than querying every model sequentially one by one. Same thing when we send queries from bash - backend time grows from 211 to 3898.

If we add time.sleep(random.random() / 2) to request function - inference goes fast with multiprocessing. It must be something with the way the server process concurrent requests. But playing with number_of_netty_threads, netty_client_threads, job_queue_size, max_request_size did not work.

Any hints or links to docs?

**Comment by dhaniram-kshirsagar** [comment]
> instance has 12 threads

Do you mean 12 cores?? or netty threads??

How many concurrent requests/model you are sending?
Also, can you attach a sampler to reproduced this.

**Comment by sermakarevich** [comment]
12 cores, yes
5 concurrent request - 1 for each model
will try to create a sampler

**Comment by sermakarevich** [comment]
@dhaniram-kshirsagar  this should work
single  request Backend response time: 400
concurrent requests Backend response time: up to 2000
```
mkdir model_store
wget https://download.pytorch.org/models/densenet161-8d451a50.pth
for i in `seq 0 10`; do
	torch-model-archiver --model-name densenet161_${i} --version 1.0 --model-file ./serve/examples/image_classifier/densenet_161/model.py --serialized-file densenet161-8d451a50.pth --export-path model_store --extra-files ./serve/examples/image_classifier/index_to_name.json --handler image_classifier
done

touch ts_config
echo "default_workers_per_model=1" >> ts_config

torchserve --start --model-store ./model_store/ --ncs --ts-config ./ts_config --models all

curl -O https://s3.amazonaws.com/model-server/inputs/kitten.jpg


for i in `seq 0 10`; do
	curl http://127.0.0.1:8080/predictions/densenet161_${i} -T kitten.jpg & done


for i in `seq 0 10`; do
	curl http://127.0.0.1:8080/predictions/densenet161_${i} -T kitten.jpg
done
```

**Comment by sermakarevich** [comment]
I think I have found the issue. Its pytorch's default parallelism. Every model worker process spawns its own subprocesses which cause conflicts. 
`MKL_NUM_THREADS=1 torchserve --start` looks like solved the issue.

**Comment by sermakarevich** [comment]
@dhaniram-kshirsagar the other strange thing here is that model's memoryUsage double its size after a request. 
model archive size: 106MB
memoryUsage after load: 175-220 MB
after an inference request (or a couple of) memoryUsage spike to ~400MB
![image](https://user-images.githubusercontent.com/9593371/85769150-04748800-b71a-11ea-8d12-090a8a5fadfe.png)
![image](https://user-images.githubusercontent.com/9593371/85769365-34239000-b71a-11ea-896f-27d0a9025627.png)

**Comment by dhaniram-kshirsagar** [comment]
@sermakarevich  As you might know, torchserve starts a python process for every model you register with it. In your case, you are registering 11 models which are equal to launching 11 python process + 1 main process. When you try to run concurrent requests against each of those 11 models, process switching results into slow response time. We tried with a 36 core machine, overall response time improves however ratio of single response and concurrent response time is almost the same!

**Comment by sermakarevich** [comment]
@dhaniram-kshirsagar thanks for the explanation - it makes sense.

**Comment by PMA65** [comment]
> @sermakarevich As you might know, torchserve starts a python process for every worker of a model registered with it. In your case, you are registering 11 models with 1 worker/model which are equal to launching 11 python processes. When you try to run concurrent requests against each of those 11 models, process switching results into slow response time. We tried with a 36 core machine, overall response time improves however response time ratio for a request and concurrent requests is almost same!

@dhaniram-kshirsagar  I'm spawning multiple worker threads for a single model (default_workers_per_model=8), I can see that GPU memory is holding 8 models but still 8 concurrent requests take 8*one response time, could you please share what config file you're using ?

**Comment by Jeozhao** [comment]
Hi @PMA65,  I meet the same trouble with you. I found that even if I use multiple works per model, the time of concurrent requests increases exponentially. Did you solve this problem ?

**Comment by PMA65** [comment]
@Jeozhao no I didn't find a solution for this, I think memory isn't the only thing that matters, you can fit 2 models in memory but that doesn't mean the GPU cores can execute those 2 with same latency as 1 model

**Comment by lschaupp** [comment]
@PMA65 @Jeozhao @sermakarevich 

I came over this thread while deploying torch serve for my own application. I thought, that I want to help anyone else who has similar issues.

My journey was a little bit different from a starting point. I actually started out with the benchmark scripts that are available in the repository.
Basically, I deployed my vision model by adapting the benchmark script and wanted to see the effect of having multiple job workers. 

Ironically, even at the beginning of using the benchmark I had almost the same performance independent of the amount of workers. However I discovered after setting: export .... , suddenly I gained 5x in performance with 5 workers. I don't understand the reason behind it -  but it is how it is ;) (related to another issue where someone recommended settings this flag)

After this I became hopeful that I could transition by writing my own scripts (hence I came across this issue) and achieve the same performance.

Initially, I used a python script with multiple processes to send the requests - however I noticed really low performance.
The encountered performance was similar to what the others wrote in this issue. (almost no gains)
I then discovered this article of benchmarking requests:

https://klaviyo.tech/a-deep-dive-into-high-performance-http-requests-for-python-engineers-2546772c50ae

A few insights: don't use multiple processes unless you have massive amounts of requests, spawning processes take a long time, it is much smarter to use async threading. (see my provided example)

`import asyncio
import aiohttp
import numpy as np

class SupportsBenchmarking:
    pass

class BenchmarkAsynchronousAsyncClient(SupportsBenchmarking):
    @staticmethod
    async def _make_request(session, url: str, file_path: str):
            # Add file to multipart writer
            with open(file_path, 'rb') as file:
                data = file

            # Send PUT request with the multipart data
            async with session.put(url, data=[]) as resp:
                if resp.status != 200:
                    print(await resp.text())
                    print("ERROR: ", resp.status)
                else:
                    response_json = await resp.text()
                    return response_json
    @classmethod
    async def _run(cls, url: str, num_requests: int, file_path: str):
        async with aiohttp.ClientSession() as session:
            tasks = [cls._make_request(session, url, file_path) for _ in range(num_requests)]
            results = await asyncio.gather(*tasks)
            return results
    @classmethod
    def run(cls, url: str, num_requests: int, file_path: str):
        results = asyncio.run(cls._run(url, num_requests, file_path))
     return results
# Example usage
url ="http://127.0.0.1:8080/predictions/mymodel"
num_requests = 1000
file_path = 'img.jpg'

import time
start = time.time()
results = BenchmarkAsynchronousAsyncClient.run(url, num_requests, file_path)
end = time.time()
print(len(results))
print("Time",end-start)
`

With this above code response, I am getting exactly the same timing result as with the benchmark code from this repository.
I wished the TorchServe team would have provided these kind of request examples in the first place for Python.
It would have saved me and everyone a lot of time debugging of what is wrong.

So, now I have a performance increase of 5-6x with using 5 workers instead of 1 - which is really great.

Also one more tip, always check if your GPU is actually fully utilized using nvidia-smi - this is a good indication whether the issue is request related or hardware bound.

**Comment by agunapal** [comment]
@lschaupp I found this explanation good for the behavior you are seeing.
https://stackoverflow.com/a/43897781

If you would like to send a PR with your benchmarking script as an example and make it configurable , please do send it.  Will be happy to review and merge.

**Comment by lschaupp** [comment]
@agunapal happy to provide an python example once I am done iterating over it -  i have made already a bunch of improvements (e.g. multiple inputs that are getting encoded).

Would not really call it benchmarking script, but rather an python request example to overcome issues that users would suffer from.

I would also recommend to add the export flag to the "Get started" guide since it massively changed things for me.

I have now actually finalized a first prototype (incorporated the above part as part of my backend), and the results translated pretty much.

**Comment by agunapal** [comment]
Hi @lschaupp I am curious if you also enabled this in your experiments. We have seen a drastic improvement with this configuration.
https://github.com/pytorch/serve/blob/master/docs/performance_guide.md#torchserve-on-cpu-
